src/plat/pc99/machine/intel-vtd.c

```diff
@@ -265,6 +294,103 @@ void vtd_handle_fault(void)
     }
 }
 
+bool_t vtd_create_irte_ioapic(word_t vector, word_t level, cpu_id_t delivery_cpu, word_t *entry_ret)
+{
+    if (x86KSnumDrhu == 0) {
+        return false;
+    }
+    /* Remapping hardware units are configured to share the interrupt-remapping table,
+     * so we just read the table address from the first unit
+     */
+    vtd_irte_t *irt_pptr = (vtd_irte_t *) paddr_to_pptr(vtd_read64(0, IRTA_REG) & IRTA_REG_IRTA_MASK);
+    /* Sanity checking whether the Interrupt Remapping Table is initialized */
+    assert(irt_pptr != paddr_to_pptr(0));
+
+    word_t entry;
+    for (entry = 0; entry <= MAX_IOAPIC && vtd_irte_ptr_get_present(&irt_pptr[entry]); entry++);
+
+    irt_pptr[entry] = vtd_irte_new(
+                          0, /* Source Validation Type */
+                          0, /* Source-id Qualifier */
+                          0, /* Source Identifier */
+                          delivery_cpu, /* Destination ID */
+                          vector, /* Vector */
+                          0, /* IRTE Mode, value 0 indicates remapped IRQ */
+                          0, /*  Delivery Mode */
+                          level, /* Trigger Mode */
+                          0, /* Redirection Hint */
+                          0, /* Destination Mode */
+                          false, /*  Fault Processing Disable */
+                          true); /* Present */
+
+    printf("IOMMU: added IRTE %lu for IOAPIC vector %lu.\n", entry, vector);
+    *entry_ret = entry;
+    return true;
+}
+
+exception_t vtd_inspect_irte_msi(word_t entry, word_t vector)
+{
+    if (x86KSnumDrhu == 0) {
+        return EXCEPTION_NONE;
+    }
+
+    if (entry < MAX_IOAPIC || entry >= N_VTD_IRTES_4K) {
+        userError("IOMMU: Invalid handle %lu for MSI vector %lu, should be between %d-%d.", entry, vector, MAX_IOAPIC,
+                  N_VTD_IRTES_4K);
+        current_syscall_error.type = seL4_RangeError;
+        current_syscall_error.rangeErrorMin = MAX_IOAPIC;
+        current_syscall_error.rangeErrorMax = N_VTD_IRTES_4K;
+        return EXCEPTION_SYSCALL_ERROR;
+    }
+
+    /* Remapping hardware units are configured to share the interrupt-remapping table,
+     * so we just read the table address from the first unit
+     */
+    vtd_irte_t *irt_pptr = (vtd_irte_t *) paddr_to_pptr(vtd_read64(0, IRTA_REG) & IRTA_REG_IRTA_MASK);
+    /* Sanity checking whether the Interrupt Remapping Table is initialized */
+    assert(irt_pptr != paddr_to_pptr(0));
+
+    if (vtd_irte_ptr_get_present(&irt_pptr[entry]) == 1) {
+        userError("IOMMU: Invalid handle %lu for MSI vector %lu, the handle is occupied by vector %llu.",
+                  entry, vector, vtd_irte_ptr_get_vector(&irt_pptr[entry]));
```

Comment:
The `%llu` gives a compile error for 32-bit x86. Not sure why 32-bit x86 is supported, but oh well.